### PR TITLE
#532: fixing foreach loops generated by set comprehension and forall expressions

### PR DESCRIPTION
**Pull request by davidcok**
This fix is just for CS and Java. JS and Go require a redesign of inheritance, which is in progress elsewhere. So submitting this partial solution but not closing the issue.

Edit: now fixed for all languages

### CODE REVIEW

File: Source/Dafny/Compiler-Csharp.cs [hunk 1]
```diff
@@ -2409,7 +2410,7 @@ private class CSharpCompilationResult
       // * CS0436 is about types in source files that conflict with imported types (caused by
       //   dynamically-generated types like Tuple0 that aren't part of the runtime, which are
       //   often in pre-compiled Dafny DLLs)
-      cp.CompilerOptions = "/debug /nowarn:0164 /nowarn:0219 /nowarn:1717 /nowarn:0162 /nowarn:0168 /nowarn:0436";
+      cp.CompilerOptions = "/debug /nowarn:0164 /nowarn:0219 /nowarn:1717 /nowarn:0162 /nowarn:0168 /nowarn:0436 /nowarn:0183";
```
Comment (RustanLeino) [inline Source/Dafny/Compiler-Csharp.cs:2413]: Update `EmitHeader()` to reflect the new no-warn.

File: Source/Dafny/Compiler-java.cs [hunk 2]
```diff
@@ -685,26 +689,31 @@ protected class ClassWriter : IClassWriter {
         if (ComplicatedTypeParameterForCompilation(argType)) {
           Error(tok, "compilation of set<TRAIT> is not supported; consider introducing a ghost", wr);
         }
+
+        if (erased) return DafnySetClass;
```
Comment (RustanLeino) [inline Source/Dafny/Compiler-java.cs:693]: Surround the then branch with curly braces. (Same comment below.)

File: Source/Dafny/Compiler-java.cs [hunk 3]
```diff
@@ -3354,18 +3363,37 @@ private class GenericArrayElementLvalue : ILvalue {
 
     protected override BlockTargetWriter CreateForeachLoop(string boundVar, Type boundVarType, out TargetWriter collectionWriter,
       TargetWriter wr, string altBoundVarName = null, Type altVarType = null, Bpl.IToken tok = null) {
+      string ty = "";
+      string tempName = "";
       if (boundVarType != null) {
-        wr.Write($"for({TypeName(boundVarType, wr, tok)} {boundVar} : ");
+        // We actually do not know the tye of the collection, which is not yet written
```
Comment (RustanLeino) [inline Source/Dafny/Compiler-java.cs:3369]: tye -> type

File: Test/git-issues/git-issue-532.dfy [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+class D extends Tr {
+  var z: int
+}
+
+method M(t: Tr)
+  modifies t
+{
+  print "t.x=", t.x, "  ";
+  var s: set<C> := set c: C | c == t;  // this line crashes for the call M(d)
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-532.dfy:25]: used to crash

File: Test/git-issues/git-issue-532.dfy [hunk 5]
```diff
@@ -0,0 +1,45 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:cs "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:js "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:go "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /compileTarget:java "%s" >> "%t"
+// RUN: %diff "%s.expect" "%t"
+// XFAIL: *
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-532.dfy:7]: Is this intended?

File: Test/git-issues/git-issue-532.dfy.expect [hunk 6]
```diff
@@ -0,0 +1,27 @@
+git-issue-532.dfy(25,19): Warning: /!\ No terms found to trigger on.
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-532.dfy.expect:1]: These no-trigger warnings can be suppressed as I show in the additional test cases.

**Comment by RustanLeino** [comment]
Regarding my additional test case, to check the constraint at run time would require the constraint to be compilable. Dafny allows these constraints to be ghost (and I'd like to keep it that way). So, the additional test case I gave should be disallowed some other way. I withdraw my suggestion to add this test case and will instead file a separate bug report for this.

File: Source/Dafny/Compiler-Csharp.cs (hunk 1 above)
Comment (davidcok) [reply Source/Dafny/Compiler-Csharp.cs:2424]: Done

File: Source/Dafny/Compiler-java.cs (hunk 3 above)
Comment (davidcok) [reply Source/Dafny/Compiler-java.cs:3342]: Fixed

File: Test/git-issues/git-issue-532.dfy (hunk 5 above)
Comment (davidcok) [reply Test/git-issues/git-issue-532.dfy]: Was. But no longer as all languages now work

File: Source/Dafny/Compiler-java.cs (hunk 2 above)
Comment (davidcok) [reply Source/Dafny/Compiler-java.cs:692]: Fixed

File: Test/git-issues/git-issue-532.dfy (hunk 4 above)
Comment (davidcok) [reply Test/git-issues/git-issue-532.dfy:24]: Fixed

File: Source/Dafny/Compiler-js.cs [hunk 7]
```diff
@@ -21,6 +21,12 @@ public JavaScriptCompiler(ErrorReporter reporter)
 
     public override string TargetLanguage => "JavaScript";
 
+    protected readonly new string DafnySetClass = "_dafny.Set";
+    protected readonly new string DafnyMultiSetClass = "_dafny.MultiSet";
+    protected readonly new string DafnySeqClass = "_dafny.ISequence";
+    protected readonly new string DafnySeqHelperClass = "_dafny.Sequence";
+    protected readonly new string DafnyMapClass = "_dafny.Map";
```
Comment (RustanLeino) [inline Source/Dafny/Compiler-js.cs:28]: Evidently, this is how the Java compiler did it. I recently saw that the new C++ compiler copied this style, and so does this JavaScript change. I think these fields should all have been declared virtual in Compiler.cs, along the lines of:

    protected virtual string DafnySetClass => "Dafny.Set";

Then, compilers that want to use different names would do:

    protected override string DafnySetClass => "dafny.DafnySet";

File: Source/Dafny/Compiler-js.cs [hunk 8]
```diff
@@ -839,19 +849,19 @@ protected class ClassWriter : IClassWriter {
         if (ComplicatedTypeParameterForCompilation(argType)) {
           Error(tok, "compilation of set<TRAIT> is not supported; consider introducing a ghost", wr);
         }
-        return DafnySetClass + "<" + TypeName(argType, wr, tok) + ">";
+        return erased ? DafnySetClass : DafnySetClass + "<" + TypeName(argType, wr, tok, erased) + ">";
```
Comment (RustanLeino) [inline Source/Dafny/Compiler-js.cs:852]: The angle brackets here in the JavaScript file is a latent bug. It's not legal JavaScript syntax, and such type parameters are not used in JavaScript in the first place. (In fact, it is rare that `TypeName` is called for JavaScript at all. This is why this bug has not been detected until now.) So, L852 should return just `DafnySetClass` regardless of the value of `erased`. This, in turn, means that the `erased` parameter is not needed in this file.

File: Source/Dafny/Compiler-Csharp.cs (hunk 1 above)
Comment (davidcok) [reply Source/Dafny/Compiler-Csharp.cs:2425]: Done

File: Source/Dafny/Compiler-js.cs (hunk 7 above)
Comment (davidcok) [reply Source/Dafny/Compiler-js.cs:28]: Fixed

File: Source/Dafny/Compiler-js.cs (hunk 8 above)
Comment (davidcok) [reply Source/Dafny/Compiler-js.cs:852]: Fixed

File: Test/git-issues/git-issue-532.dfy.expect (hunk 6 above)
Comment (davidcok) [reply Test/git-issues/git-issue-532.dfy.expect]: Fixed as described

**Pull request merged**